smithy-model/src/main/java/software/amazon/smithy/model/shapes/ShapeId.java

```diff
@@ -131,21 +131,25 @@ private static int parseIdentifier(CharSequence identifier, int offset) {
             return -1;
         }
 
-        // Parse the required identifier_start production.
+        // Parse the required IdentifierStart production.
         char startingChar = identifier.charAt(offset);
         if (startingChar == '_') {
-            while (identifier.charAt(offset) == '_') {
+            while (offset < identifier.length() && identifier.charAt(offset) == '_') {
                 offset++;
             }
-            if (!ParserUtils.isValidIdentifierCharacter(identifier.charAt(offset))) {
+            if (offset == identifier.length()) {
+                return -1;
+            }
+            if (!ParserUtils.isAlphabetic(identifier.charAt(offset))
```

Comment:
super minor nit, given this method is called thousands of times: move the current character to a local variable to avoid bounds checks multiple times with charAt.

e.g.:
```
            char current = identifier.charAt(offset);
            if (!ParserUtils.isAlphabetic(current) && !ParserUtils.isDigit(current)) {
                return -1;
            }
```